Adding ability to automatically set DAG to off after X times it failed sequentially

pateash opened the pull request:
closes: #36239

## Description
This PR adds ability to automatically turn off DAG after X number of consecutive failures.

There are 2 ways to configure this
1. in DAG() using **max_consecutive_failed_dag_runs** parameter this will override any config (as below)
2. Using core config inside airflow.cfg or variable AIRFLOW__CORE_MAX_CONSECUTIVE_FAILED_DAG_RUNS_PER_DAG 
environment variable.

pateash committed:
Add max_failure_runs field to DagModel

A new column 'max_failure_runs' has been added to the 'dag' table in the database and the corresponding model 'DagModel'. This column is used to store the maximum number of allowed failure runs per DAG. If the limit is reached, the scheduler will disable the DAG. Default configuration for this field has been added to the 'airflow' config file.

pateash committed:
Implement max_failure_runs option for DAGs

A new attribute 'max_failure_runs' has been added to 'DagModel'. It specifies the maximum number of consecutive failed DAG runs. Once the limit is exceeded, the scheduler will disable the corresponding DAG. An invalid value exception has been added ensuring 'max_failure_runs' is not less than zero.

pateash committed:
Add method to check last N DAG runs' failure status

A new method, check_last_N_dagruns_failed, has been added to the dagrun model. This method checks if the last N DagRuns have all failed. Its main purpose is to introduce checks for consecutive failures, allowing the scheduling system to handle consecutive DagRun failures more effectively.

pateash committed:
Add 'max_failure_runs' to various airflow files

The 'max_failure_runs' field has been added to DagModel schema, pydantic dagger file, CLI dag command, and connection schema. This new field is set to store maximum allowed failure runs for each DAG in the database, helping to automate the process of disabling DAG when max runs are reached.

pateash committed:
Refactor 'check_last_N_dagruns_failed' method and add 'max_failure_runs' check

Renamed the 'check_last_N_dagruns_failed' method to '_check_last_N_dagruns_failed' for indicating it as a private method. Additionally, a new functionality has been added to check for 'max_failure_runs'. If the number of consecutive DAG failures exceeds this limit, the DAG gets automatically marked as paused.

pateash committed:
Add auto-pausing of DAG on exceeding 'max_failure_runs'

Refactored the '_check_last_N_dagruns_failed' method and integrated a check for 'max_failure_runs'. If the number of consecutive failure runs of the DAG exceeds this limit, the DAG is now automatically set to a paused status to prevent further failures.

dirrao commented on airflow/models/dagrun.py:
```diff
@@ -498,6 +498,37 @@ def fetch_task_instances(
+            .limit(number_of_dag_runs).all()
+
+        """ Marking dag as paused, if needed"""
+        toBePaused = all(dag_run.state == DagRunState.FAILED for dag_run in dag_runs)
```
To ensure the correct behavior, we need to check the count of failures as well.

pateash replied:
We are checking last N dagruns and checking if all the failed,
this will make sure we have the required count.
please highlight if I am missing some edge case here.

dirrao replied:
I am referring to the limit query might give less records. So we need cross check the failed count.

pateash replied:
ohh, yeah
thanks for pointing it out.
fixed this edgecase.

dirrao reviewed (changes requested):
Great work. Can you add the test cases for this feature?

BasPH commented on airflow/config_templates/config.yml:
```diff
@@ -116,6 +116,15 @@ core:
       type: string
       example: ~
       default: "16"
+    max_failure_runs_per_dag:
```
```suggestion
    max_consecutive_failed_dag_runs:
```

Would add "consecutive" for clarity

BasPH commented on airflow/config_templates/config.yml:
```diff
@@ -116,6 +116,15 @@ core:
+      description: |
+        The maximum number of consecutive DAG failures before DAG is turned off. The scheduler will disable the DAG
+        if it reaches the limit. This is configurable at the DAG level with ``max_failure_runs``,
+        which is defaulted as ``max_failure_runs_per_dag``.
```
```suggestion
        The maximum number of consecutive failed DAG runs before a DAG is automatically paused. This is also configurable per DAG with the argument ``max_consecutive_failed_dag_runs``.
```

WDYT of having the same name for the global config and per-DAG config for clarity? This avoids config name sprawl.

BasPH commented on airflow/models/dag.py:
```diff
@@ -611,12 +614,18 @@ def __init__(
+        if self.max_failure_runs is not None and self.max_failure_runs < 0:
+            raise AirflowException(
+                f"Invalid max_failure_runs: {str(self.max_failure_runs)}"
+                f"requires max_failure_runs >= 0"
```
```suggestion
                f"Invalid max_failure_runs: {self.max_failure_runs}. "
                "Requires max_failure_runs >= 0"
```

Don't believe string conversion is required here. Also adding whitespace between sentences + remove unnecessary f-string.

BasPH commented on airflow/models/dagrun.py:
```diff
@@ -498,6 +498,37 @@ def fetch_task_instances(
         return session.scalars(tis).all()
 
+    @internal_api_call
+    def _check_last_N_dagruns_failed(
```
```suggestion
    def _check_last_n_dagruns_failed(
```

Just keep everything lowercase

BasPH commented on airflow/models/dagrun.py:
```diff
@@ -498,6 +498,37 @@ def fetch_task_instances(
+            .order_by(DagRun.execution_date.desc()) \
+            .limit(number_of_dag_runs).all()
+
+        """ Marking dag as paused, if needed"""
```
```suggestion
        """ Marking DAG as paused, if needed"""
```

BasPH commented on airflow/models/dagrun.py:
```diff
@@ -498,6 +498,37 @@ def fetch_task_instances(
+            .limit(number_of_dag_runs).all()
+
+        """ Marking dag as paused, if needed"""
+        toBePaused = all(dag_run.state == DagRunState.FAILED for dag_run in dag_runs)
```
```suggestion
        to_be_paused = all(dag_run.state == DagRunState.FAILED for dag_run in dag_runs)
```

BasPH commented on airflow/models/dagrun.py:
```diff
@@ -498,6 +498,37 @@ def fetch_task_instances(
+        toBePaused = all(dag_run.state == DagRunState.FAILED for dag_run in dag_runs)
+        if toBePaused:
+            from airflow.models.dag import DagModel
+            self.log.info("Marking Dag %s paused", self.dag_id)
```
```suggestion
            self.log.info("Pausing DAG %s because last %s DAG runs failed.", self.dag_id, number_of_dag_runs)
```

Simplify wording + add rationale for an event, this helps debugging.

BasPH commented on airflow/models/dagrun.py:
```diff
@@ -498,6 +498,37 @@ def fetch_task_instances(
+    def _check_last_N_dagruns_failed(
+        self,
+        dag_id: str,
+        number_of_dag_runs: int,
```
Can we just give this argument the same name as the config to avoid variable naming sprawl? E.g. `max_consecutive_failed_dag_runs`, this helps code readability.

BasPH commented on airflow/models/dagrun.py:
```diff
@@ -498,6 +498,37 @@ def fetch_task_instances(
+                .execution_options(synchronize_session="fetch")
+            )
+        else:
+            self.log.info("Limit of consecutive dag failed is not reached, Dag %s is not being paused.",
```
```suggestion
            self.log.info("Limit of consecutive failed DAG runs is not reached, DAG %s will not be paused.",
```

BasPH commented on airflow/models/dagrun.py:
```diff
@@ -724,6 +755,13 @@ def recalculate(self) -> _UnfinishedStates:
                     msg="task_failure",
                 )
 
+            # checking if the max_failure_runs has been provided and last consecutivate failures are more
```
```suggestion
            # Check if the max_failure_runs has been provided and last consecutive failures are more
```

BasPH commented on airflow/models/dagrun.py:
```diff
@@ -724,6 +755,13 @@ def recalculate(self) -> _UnfinishedStates:
                 )
 
+            # checking if the max_failure_runs has been provided and last consecutivate failures are more
+            # than this number if so we have to mark this dag as off
```
```suggestion
            # than this number if so we have to pause this DAG
```

Lets use the same wording for everything for readability, so instead of "marking off" -> "pausing".

BasPH commented on airflow/models/dagrun.py:
```diff
@@ -724,6 +755,13 @@ def recalculate(self) -> _UnfinishedStates:
+            # checking if the max_failure_runs has been provided and last consecutivate failures are more
+            # than this number if so we have to mark this dag as off
+            if bool(dag.max_failure_runs):
+                self.log.info("Checking consecutive failed dags for %s, limit is %s", self.dag_id,
```
```suggestion
                self.log.info("Checking consecutive failed DAG runs for DAG %s, limit is %s", self.dag_id,
```

BasPH reviewed (commented):
I think this would be a helpful feature. Left a few comments around naming conventions

pateash committed:
Rename 'max_failure_runs' to 'max_consecutive_failed_dag_runs'

The 'max_failure_runs' parameter across multiple files was renamed to 'max_consecutive_failed_dag_runs' to better represent its functionality. This change improves code readability, and reflects that this parameter is used to limit the number of consecutive failed DAG runs, after which the DAG will be disabled.

pateash committed:
Refactor method and variable names, enhance log messages in dagrun.py

Method '_check_last_N_dagruns_failed' was renamed to '_check_last_n_dagruns_failed' and variable 'toBePaused', to 'to_be_paused' for better readability and adherence to Python naming conventions. Additionally, log messages were enhanced to provide clearer and more informative feedback about the DAG runs' status, specifically in the context of maximum consecutive failed DAG runs.

pateash committed:
Corrected downgrade method's docstring in migration file

The docstring of the downgrade method in the '0134_2_8_1_adding_adding_max_failure_runs_column_.py' migration file was corrected. It previously had 'Adding' mentioned twice unnecessarily. This revision improves readability and clarity of the operation being performed by the method.

pateash committed:
Refactor variable names and update descriptions in Airflow models

Variable 'number_of_dag_runs' is renamed to 'max_consecutive_failed_dag_runs' for better clarity in dagrun.py and dag.py. Additionally, the configuration description of 'max_consecutive_failed_dag_runs_per_dag' in config.yml is updated to reflects the purpose better: the automatic pausing of DAG runs after a specified number of consecutive failures.

pateash committed:
Refined logic for pausing DAG upon failures

The logic to pause a DAG run in the Airflow model is refined to check that the number of DAG run failures is equal to or exceeds the set maximum consecutive failed DAG runs. This change better identifies a DAG that needs to be paused, ensuring that the system efficiently handles failure situations.

pateash committed:
Add 'max_consecutive_failed_dag_runs' to DAG schema

In the DAG schema, 'max_consecutive_failed_dag_runs' has been added. This attribute determines the threshold for how many consecutive failures will result in the DAG being paused. This will improve error handling, ensuring DAGs are paused promptly if they are consistently failing.

pateash committed:
Add default configuration for `max_consecutive_failed_dag_runs`

In this commit, we added a default value for `max_consecutive_failed_dag_runs` in `airflow/config_templates/config.yml`. It means that if not specified, the dags are not paused by default. Also, the readability and clarity of the operation being performed by the downgrade method in the migration file were improved by correcting the docstring description.

pateash committed:
Update comment in `dagrun.py`

Updated the comment in `dagrun.py` to indicate that the `max_consecutive_failed_dag_runs` configuration has to be provided and not 0. This makes the purpose of the check more explicit and hence improves code readability.

pateash committed:
Add tracing to test_dag and new DAG pause test case

Integrated `pydevd_pycharm` package to `tests/models/test_dag.py` for enhanced debugging. Also, added a new test case to validate the 'pause' status of a DAG upon consecutive failures. This test checks whether a DAG pauses after hitting the limit for maximum consecutive failed dag runs.

pateash committed:
Refactor test_dag and add new test for DAG pause behavior

Refactored `test_dag.py` by abstracting repeated code into `add_failed_dag_run` function. This makes the tests more readable and maintainable. Furthermore, a new test was added to verify the 'pause' behavior of a DAG after a limit of consecutive failures has been reached. This ensures that the DAG behaves as expected under failure conditions.

pateash committed:
Refactor test_dag and add new test for DAG pause behavior

Refactored `test_dag.py` by abstracting repeated code into `add_failed_dag_run` function. This makes the tests more readable and maintainable. Furthermore, a new test was added to verify the 'pause' behavior of a DAG after a limit of consecutive failures has been reached. This ensures that the DAG behaves as expected under failure conditions.

pateash committed:
Remove debug code from test_dag.py

The imported pydevd_pycharm debugging code has been removed from `test_dag.py`. This code is unnecessary for production and may create potential security issues, hence the removal.

pateash committed:
Add max_consecutive_failed_dag_runs to test data

The max_consecutive_failed_dag_runs field has been added in several locations within the test data in test_dag_endpoint.py. This additional field will provide better test coverage for scenarios where consecutive DAG runs have failed.

pateash committed:
Refactor code for readability and standards compliance

Refactored the code in several files including dagrun.py and dag.py to improve readability and adhere to coding standards. Changes include reformatting complex statements and repositioning multiline arguments. Also updated corresponding documentation and test data to reflect these changes.

pateash committed:
Update version in migration script filename and contents

Renamed the migration script from version '2.8.1' to '2.9.0' and updated the 'airflow_version' variable within the file to '2.9.0'. The changes correspond to the new airflow version and ensure that the migration process is consistent with this version.

eladkal commented on airflow/api_connexion/schemas/dag_schema.py:
```diff
@@ -69,6 +69,7 @@ class Meta:
     tags = fields.List(fields.Nested(DagTagSchema), dump_only=True)
     max_active_tasks = auto_field(dump_only=True)
     max_active_runs = auto_field(dump_only=True)
+    max_consecutive_failed_dag_runs = auto_field(dump_only=True)
```
I wonder about the name. `max_consecutive_failed_dag_runs` might be interpreted as not allowing scheduling after the X failures but this is not really a hard limit - user can override it by enabling the dag. For the other max properties user can't really override them.

pateash replied:
@eladkal,
with current logic,
If a user wants they can manually enable the dag, but if it fails again, then the dag will automatically be disabled, as we are counting last N failed which will again be true.
if user wants to restart counting again, they will have add to add a non-failed dag_run for now.

pateash replied:
for naming, 
yeah its a good point.
happy to rename it , please suggest if we can come up with better naming for messaging.

eladkal replied:
I need to think about the name.

> if user wants to restart counting again, they will have add to add a non-failed dag_run for now

That might be problematic however I think we can avoid deciding right now.
We can mark this feature as experimintal thus it allows the flexability of changing the behavior if needed.

amoghrajesh replied:
I thought a bit about this and what about: `max_allowed_dag_failures_in_row` or just `max_dag_failures_in_row`?

eladkal reviewed (commented):
Looks good!

I have a question about edge case.
Say that I set max_consecutive_failed_dag_runs=3.
After 3 failures the DAG is set to off, if I set it to on manually and the next run also failed what will happen? will it also set the dag to off or the count restart?

Basically the question is if we should count only the failed dags or also take into account events that might rest the count like setting from off to on.

pateash committed:
tests: fixed

pateash committed:
fixed: pre-commits

eladkal commented on airflow/config_templates/config.yml:
```diff
@@ -116,6 +116,17 @@ core:
+        This is also configurable per DAG level with ``max_consecutive_failed_dag_runs``,
+        which is defaulted as ``max_consecutive_failed_dag_runs_per_dag``.
+        If not specified, then the value is considered as None,
+        meaning that the dags are never paused out by default.
```
Lets mention this as experimental feature.

We need also:
1. News fragment entry for feature (that gives 2-3 sentences about this capability and explain this is experimental one)
2. Doc entry that explain this feature and explain this is experimental one


After these LGTM

pateash committed:
fixed: migrations added

pateash committed:
fixed: migrations added

uranusjr commented on airflow/config_templates/config.yml:
```diff
@@ -321,12 +332,12 @@ core:
       description: |
         The default task execution_timeout value for the operators. Expected an integer value to
-        be passed into timedelta as seconds. If not specified, then the value is considered as None,
+        be passed into timedelta as seconds. If not specified, then the value is considered as 0,
```
Is this related?

pateash replied:
No, fixed.

uranusjr commented on airflow/models/dag.py:
```diff
@@ -454,6 +456,7 @@ def __init__(
         concurrency: int | None = None,
         max_active_tasks: int = airflow_conf.getint("core", "max_active_tasks_per_dag"),
         max_active_runs: int = airflow_conf.getint("core", "max_active_runs_per_dag"),
+        max_consecutive_failed_dag_runs: int | None = None,
```
It’s better to call `getint` here as the default instead. The None type can go away too.

uranusjr commented on airflow/models/dag.py:
```diff
@@ -615,6 +618,16 @@ def __init__(
+            )
+        if self.max_consecutive_failed_dag_runs < 0:
+            raise AirflowException(
+                f"Invalid max_consecutive_failed_dag_runs: {str(self.max_consecutive_failed_dag_runs)}."
```
```suggestion
                f"Invalid max_consecutive_failed_dag_runs: {self.max_consecutive_failed_dag_runs}."
```

uranusjr commented on airflow/models/dag.py:
```diff
@@ -3574,6 +3588,7 @@ class DagModel(Base):
 
     max_active_tasks = Column(Integer, nullable=False)
     max_active_runs = Column(Integer, nullable=True)
+    max_consecutive_failed_dag_runs = Column(Integer, nullable=True)
```
Don’t need to be nullable.

uranusjr commented on airflow/models/dag.py:
```diff
@@ -3922,6 +3942,7 @@ def dag(
     concurrency: int | None = None,
     max_active_tasks: int = airflow_conf.getint("core", "max_active_tasks_per_dag"),
     max_active_runs: int = airflow_conf.getint("core", "max_active_runs_per_dag"),
+    max_consecutive_failed_dag_runs: int | None = None,
```
Same

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -561,6 +561,45 @@ def fetch_task_instances(
+        """ Marking dag as paused, if needed"""
+        to_be_paused = len(dag_runs) >= max_consecutive_failed_dag_runs and all(
+            dag_run.state == DagRunState.FAILED for dag_run in dag_runs
+        )
```
```suggestion
        dag_run_states = session.scalars(
            select(DagRun.state)
            .where(DagRun.dag_id == dag_id)
            .order_by(DagRun.execution_date.desc())
            .limit(max_consecutive_failed_dag_runs)
        ).all()

        to_be_paused = len(dag_runs) >= max_consecutive_failed_dag_runs and all(
            state == DagRunState.FAILED for state in dag_run_states
        )
```

It’s likely possible to not even do the `.all()` part, but let’s keep this simple.

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -561,6 +561,45 @@ def fetch_task_instances(
+            self.log.info(
+                "Limit of consecutive DAG failed dag runs is not reached, DAG %s will not be paused.",
+                self.dag_id,
+            )
```
I don’t think this should be `info`; `debug` is probably good enough.

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -561,6 +561,45 @@ def fetch_task_instances(
+        if to_be_paused:
+            from airflow.models.dag import DagModel
+
+            self.log.warning(
```
This should be `info` since the operation is not inherently a problem in the scheduler.

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -787,6 +826,16 @@ def recalculate(self) -> _UnfinishedStates:
 
+            # Check if the max_consecutive_failed_dag_runs has been provided and not 0
+            # and last consecutive failures are more
+            if bool(dag.max_consecutive_failed_dag_runs):
```
```suggestion
            if dag.max_consecutive_failed_dag_runs > 0:
```

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -787,6 +826,16 @@ def recalculate(self) -> _UnfinishedStates:
+            # Check if the max_consecutive_failed_dag_runs has been provided and not 0
+            # and last consecutive failures are more
+            if bool(dag.max_consecutive_failed_dag_runs):
+                self.log.info(
```
Also shouldn’t be `info` IMO.

pateash replied:
sure, marking this as debug

pateash committed:
fixed: migrations added

pateash committed:
fixed: migrations added

pateash committed:
fixes: pr comments

pateash committed:
fixes: pr comments

eladkal commented on airflow/config_templates/config.yml:
```diff
@@ -116,6 +116,17 @@ core:
       default: "16"
+    max_consecutive_failed_dag_runs_per_dag:
+      description: |
+        The maximum number of consecutive DAG failures before DAG is automatically paused.
```
```suggestion
        (experimental) The maximum number of consecutive DAG failures before DAG is automatically paused.
```

pateash committed:
Update airflow/config_templates/config.yml

Co-authored-by: Elad Kalif <45845474+eladkal@users.noreply.github.com>

pateash committed:
fixes: experimental added

eladkal commented on airflow/config_templates/config.yml:
```diff
@@ -326,7 +337,7 @@ core:
       type: integer
       example: ~
-      default: ""
+      default: None
```
Why did you change this?
It causes tests to fail
`airflow.exceptions.AirflowConfigException: Failed to convert value to int. Please check "default_task_execution_timeout" key in "core" section. Current value: "None".`

pateash replied:
this is unrelated, reverting back
looks like a merge error.

pateash committed:
fixes: reverting unrelated changes

pateash committed:
fixes: pr comments

pateash committed:
fixes: pr comments

pateash committed:
fixes: pr comments

pateash committed:
fixes: pr comments

pateash committed:
fixes: docs added

pateash committed:
fixed: pre-commits

pateash committed:
fixed: pre-commits

eladkal commented on airflow/config_templates/config.yml:
```diff
@@ -116,6 +116,17 @@ core:
+        which is defaulted as ``max_consecutive_failed_dag_runs_per_dag``.
+        If not specified, then the value is considered as 0,
+        meaning that the dags are never paused out by default.
+      version_added: ~
```
```suggestion
      version_added: 2.9.0
```

eladkal commented on airflow/models/dag.py:
```diff
@@ -364,6 +364,8 @@ class DAG(LoggingMixin):
     :param max_active_runs: maximum number of active DAG runs, beyond this
         number of DAG runs in a running state, the scheduler won't create
         new active DAG runs
+    :param max_consecutive_failed_dag_runs:(experimental) maximum number of consecutive failed DAG runs,
```
```suggestion
    :param max_consecutive_failed_dag_runs: (experimental) maximum number of consecutive failed DAG runs,
```

eladkal replied:
cc @pateash this missing space is the cause of the failing doc build

eladkal committed:
Update airflow/config_templates/config.yml

eladkal committed:
Update airflow/models/dag.py

eladkal reviewed (approved):
LGTM
Will merge once CI is green

pateash committed:
fixed: scalars giving wrong state
